core/src/test/java/hudson/model/FingerprintCleanupThreadTest.java

```diff
@@ -90,6 +94,32 @@ public void testIOExceptionOnLoad() throws IOException {
         assertTrue("Should have logged IOException.", logOutput.contains("ERROR: Failed to process"));
     }
 
+    @Test
+    public void testBlockingFacetBlocksDeletion() throws IOException {
+        createFolderStructure();
+        TestTaskListener testTaskListener = new TestTaskListener();
+        Fingerprint fp = new TestFingerprint(false);
+        fp.facets.setOwner(Saveable.NOOP);
+        TestFingperprintFacet facet = new TestFingperprintFacet(fp, System.currentTimeMillis(), true);
+        fp.facets.add(facet);
+        FingerprintCleanupThread cleanupThread = new TestFingerprintCleanupThread(fp);
+        cleanupThread.execute(testTaskListener);
+        assertTrue("Should have deleted obsolete file.", fpFile.toFile().exists());
```

Comment:
Is there a race condition here where the cleanup thread may not yet have completed the deletion of the fingerprint file?

Also, is it worth considering using a hamcrest assertion rather than assertTrue?

```suggestion
        assertThat(fpFile.toFile()., is(aReadableFile()));
```